File: adventofcode2023-go/pkg/day1.go
package pkg

import (
	"strings"

	"com.github/lprakashv/adventofcode2023/input"
)

func CalibrationSum1() int {
	sum := 0

	lines := strings.Split(input.Day1Input, "\n")
	for _, line := range lines {
		var digits []int

		for _, ch := range line {
			if ch >= '0' && ch <= '9' {
				digits = append(digits, int(ch-'0'))
			}
		}

		if len(digits) > 0 {
			sum += digits[0]*10 + digits[len(digits)-1]
		}
	}

	return sum
}

func replaceDigitWords(s string) string {
	replacements := []string{
		"zero", "z0o",
		"one", "o1e",
		"two", "t2o",
		"three", "t3e",
		"four", "f4r",
		"five", "f5e",
		"six", "s6x",
		"seven", "s7n",
		"eight", "e8t",
		"nine", "n9e",
	}
	for i := 1; i < len(replacements); i += 2 {
		s = strings.ReplaceAll(s, replacements[i-1], replacements[i])
	}
	return s
}

func CalibrationSum2() int {
	sum := 0

	lines := strings.Split(input.Day1Input, "\n")
	for _, line := range lines {
		replaced := replaceDigitWords(line)

		var digits []int

		for _, ch := range replaced {
			if ch >= '0' && ch <= '9' {
				digits = append(digits, int(ch-'0'))
			}
		}

		if len(digits) > 0 {
			sum += digits[0]*10 + digits[len(digits)-1]
		}
	}
	return sum
}

File: adventofcode2023-clojure/src/lprakashv/adventofcode2023/day1.clj
(ns lprakashv.adventofcode2023.day1
  (:gen-class)
  (:require clojure.java.io
            [clojure.string :as s]))

(defn- positive-numbers
  ([] (positive-numbers 1))
  ([n] (lazy-seq (cons n (positive-numbers (inc n))))))

(defn- replace-worded-nums-to-digits [string]
  (-> string
      (s/replace #"zero" "z0o")
      (s/replace #"one" "o1e")
      (s/replace #"two" "t2o")
      (s/replace #"three" "t3e")
      (s/replace #"four" "f4r")
      (s/replace #"five" "f5e")
      (s/replace #"six" "s6x")
      (s/replace #"seven" "s7n")
      (s/replace #"eight" "e8t")
      (s/replace #"nine" "n9e")))

(defn calibration-sum-1
  []
  (->> (slurp (.getFile (clojure.java.io/resource "day1_input.txt")))
       (s/split-lines)
       (map
        (fn [line]
          (->> (vec line)
               (filter #(Character/isDigit %))
               ((fn [char-list] [(first char-list) (last char-list)]))
               (apply str)
               (Integer/parseInt))))
       (reduce +)))

(defn calibration-sum-2
  []
  (->> (slurp (.getFile (clojure.java.io/resource "day1_input.txt")))
       (s/split-lines)
       (map
        (fn [line]
          (->> line
               (replace-worded-nums-to-digits)
               (vec)
               (filter #(Character/isDigit %))
               ((fn [char-list] [(first char-list) (last char-list)]))
               (apply str)
               (Integer/parseInt))))
       (reduce +)))